### ISSUE TITLE
Question: libuv with nats-c for async single-threaded design

### ISSUE THREAD

**Issue by pananton** [nats-io/nats.c#299]
Hello. I have one more question regarding natsc-libuv integration. In my project I have several microservices each operating on single thread and using completly async berkley sockets under the hood. This design is sufficient for  me and leads to more simple and error-prone code, because it requires almost no synchronization.  Now I want to incorporate nats, but continue to use single thread approach. Latency is the highest priority, so I disabled buffering with send asap option. I was looking into following solutions:
1. Do not use libuv with natsc, rather use non-blocking NextMsg in my event loop for polling for new messages and then calling corresponding cb - but in that case publish-related nats calls (as far as I'm concerned) will use sync send calls. Is it true?
2. Because I worry that sync sends can harm my high load publishers, I look into libuv with natsc usage. Looks like in that case publish-related calls will be async (natsc will buffer data and return immediately, then epoll in libuv indicates socket is ready and send takes place in libuv loop thread) The problem is I've only found async subscribers example for libuv (when OnMsg cb is specified during subscription, NextMsg is not used). And natsc documetation tells that async subscriptions callbacks are called from separate threads. Is it true even in case when libuv is used? Is there a possibility to call subscription callbacks from libuv loop thread?

**Comment by kozlovic** [comment]
> Do not use libuv with natsc, rather use non-blocking NextMsg in my event loop for polling for new messages and then calling corresponding cb - 

Not sure I fully understand the question. But, you could indeed create a synchronous subscription (natsConnection_SubscribeSync()) and whenever you want to check for available messages but don't block for it, then call natsSubscription_NextMsg() with 0 as a timeout and this call will then either return the first message that is in the internal list or timeout error if not.

> but in that case publish-related nats calls (as far as I'm concerned) will use sync send calls. Is it true?

You mean if you don't use libuv? If you publish without the send asap option, then messages are buffered and flushed by a dedicated thread. If you use the send asap option, each message will be published in place.

> Because I worry that sync sends can harm my high load publishers, I look into libuv with natsc usage. Looks like in that case publish-related calls will be async (natsc will buffer data and return immediately, then epoll in libuv indicates socket is ready and send takes place in libuv loop thread)

Correct. With libuv, the Publish() calls are just adding the message to the buffer that when libuv event loop notice that there is a write event, will call the registered function back into NATS C library and do the actual send. The send asap option will cause each Publish() call to ensure that a write event is registered with the event loop, but I think it would be ok without.

> The problem is I've only found async subscribers example for libuv (when OnMsg cb is specified during subscription, NextMsg is not used). And natsc documetation tells that async subscriptions callbacks are called from separate threads. Is it true even in case when libuv is used? 

Yes, async subscriptions are dispatched from their own thread, alternatively by a library's thread pool, but not from the libuv event pool.

> Is there a possibility to call subscription callbacks from libuv loop thread?

See answer to point 1. You could always invoke NextMsg() with timeout 0.

**Comment by pananton** [comment]
Suppose that send asap is used and kernel send buffer is filled. If natsc publish will be called when libuv is used I think it will not block, because libuv itself will trigger write only when kernel is able to send more data. Will piblish block in the same situation (send asap, kernel send buffer is completely filled) but without libuv?

**Comment by kozlovic** [comment]
When using libuv, the Publish() call will never be done in place, that is, the call won't block. The event loop will cause the buffer to be flushed. So to be clear, wherever in your code you would have:
```
natsConnection_Publish(nc, "foo", someData, someDataLen);
```
this call will never block, regardless the size of the buffer..

**Comment by pananton** [comment]
Yep, this is clear to me. But without libuv can publish block when network is congested and no space remains in kernel send buffer? Because socket send will block in that case if socket is not set to non-blocking mode.

**Comment by kozlovic** [comment]
Oh sorry, I thought you were still asking about libuv. So without libuv, yes, a Publish() call that would fill the underlying buffer or due to send-asap will try to send in place would cause the Publish() call to block, unless you are natsOptions_SetWriteDeadline() (see https://nats-io.github.io/nats.c/group__opts_group.html#gaafda5fb8d9ee3a17bd475ceaecfac218), in that case socket is set to non blocking mode. But then if the write fails due to this timeout the connection is marked as closed..

**Comment by pananton** [comment]
Thanks for your answers!

**Comment by pananton** [comment]
Sorry, one more question regarding write deadline. Setting this option still makes publish call to block until timeout is not expired? I.e. write deadline only provides guarantees not to block indefinetly in publish call but does not make it return immediately while trying to deliver message in background?

**Comment by kozlovic** [comment]
> Setting this option still makes publish call to block until timeout is not expired? I.e. write deadline only provides guarantees not to block indefinetly in publish call but does not make it return immediately while trying to deliver message in background?

Correct.

> And also want to ask you about connection draining. I use function that allows to specify drain timeout and need to know whether drain was succesful or timeout expired and connection was not drained (is it hardclosed in that case?) Thanks in advance!

Yes, the connection will be closed and an error (timeout) will be submitted to the async error handler and available as natsConnection_GetLastError().

**Comment by pananton** [comment]
Ivan, sorry for bothering you, but I've found some strange problem with natsConnection_DrainTimeout function for which I can't find explanation in documentation. Documentation explicitly states that this function never blocks. In my app I use small drain timeout (about 5 seconds) and uv loop. I'm calling uv_run with UV_RUN_NOWAIT which means that it only processes pending events and returns. Later I call natsConnection_DrainTimeout from the same thread where  uv_run(UV_RUN_NOWAIT) is called and the function is blocked for 60 seconds - value which is used as connection timeout in my app. Is it my app bug or some issues in library? If I use uv_run(UV_RUN_DEFAULT) instead everything works fine.

**Comment by kozlovic** [comment]
@pananton Unfortunately, it looks like Drain[Timeout]() is actually blocking since it does internally some natsConnection_Flush() calls, which would block in the context of libuv since the call waits in place for a reply that is not going to come until libuv runs the next loop iteration. 
I think that I should have made Drain() always execute the internal calls to flush the subscriptions in a different thread, but that would not solve the problem that Drain() will always fail in the context of libuv (or other event loops) because of the internal use of request/reply (flush is sending a PING and waiting synchronously for a PONG). Even if I were to remove the Flush() that I am doing per subscription (which may not be needed), there is one for the publish calls. So this one would always timeout (right now it is 60 seconds, but I should be using the Drain timeout value) and therefore always produce a "unable to flush" error and always take the drain timeout to complete. So really you can't use that with event loops.

I have a section in the README about the event loop where I list some of the APIs that should not be used with event loops for that same reason: https://github.com/nats-io/nats.c#using-an-event-loop-library. I could add Drain[Timeout]() there..

**Comment by pananton** [comment]
I think it will be good. Thanks!

**Comment by pananton** [comment]
Hello, Ivan! Can you please tell me whether it is allowed by library design to use several uv loops in one app ? In your library examples you use uv_default_loop, but I've tried to use uv_loop_init and create several uv_loop_t structures each launched on separate thread (looks like it is allowed by libuv itself). Now I'm getting segmentation fault in natsConnection_DrainTimeout if I use several loops. This can be my bug, but maybe it is a nats.c library limitation? Also if I use two uv_loop_t but one only for publishing another for subscribing everything looks ok, at least I can't reproduce failure.

**Comment by pananton** [comment]
I've spent more time discovering the problem and can tell that natsConnection_DrainTimeout is not the source of problem - my app continues crashing in different places, so I think it's some problems with synchronization inside my code. As I've wrote earlier, I use separate threads for several uv loops in RUN_DEFAULT mode and global thread pool (set in natsOptions) for subscription messages dispatching.  I've tried distinct configuration - when I use 1 subscriber loop (i.e. connection, attached to it only subscribes and does not publish) and large number of publisher loops (10-20 connections that perform only publishing) everything works fine. But if I launch several subscribers I get a crash from time to time. Just want to be sure, that I'm not making wrong assumption that nats.c allows many simultaneous uv loops. Of course, each on its own thread and initialized with natsLibuv_SetThreadLocalLoop / natsOptions_SetEventLoop .

**Comment by kozlovic** [comment]
@pananton As long libuv allows you to do so, then it should be ok, again assuming that you set the event loop per connection.
Note that as of now, natsConnection_DrainTimeout will not work well for event loops since the call may block doing an internal flush. I may try to address that by having a different thread doing the actual work.

**Comment by pananton** [comment]
Yeah, I have a special thread for draining/flushing as you said. Ok, will try to find out the reason, thanks. Btw, do you have any tests for nats.c with libuv ?

**Comment by kozlovic** [comment]
I provide examples for both libuv and libevent that can be optionally compiled. I always try those before a release, but there is no actual unit tests in test.c.

**Comment by pananton** [comment]
Ivan, can you please also explain whether it is necessary to call nats_ReleaseThreadMemory in threads that run uv loop? I do not call anything except uv_run(RUN_DEFAULT) in them - should I still call mentioned above function when uv_run returns and thread exits? And if I call release function in thread that did not call any nats function (for example, it was stopped and did not have time to perform any work) - will it be an error?

**Comment by pananton** [comment]
Also can you tell me whether natsSubscription_IsValid returning false provides strong guarantees that onMsg callback will not be called more for the subscription? I've read in documentation that after unsubscribe dispatched callbacks can still be called if already dispatched, so I'm interesting can I use return value of natsSubscription_IsValid to understand, that no more callbacks remaining. 

Btw, can you add some callback to discover when subscription is closed and no more messages will be received. Without such a callback I have troubles with parameters passed as onMsg closure - if I create them dynamically it's hard to determine when to free them (or maybe I'm missing some possibility).

Thanks in advance!

**Comment by kozlovic** [comment]
> can you please also explain whether it is necessary to call nats_ReleaseThreadMemory in threads that run uv loop? I do not call anything except uv_run(RUN_DEFAULT) in them - should I still call mentioned above function when uv_run returns and thread exits? 

On top of the documentation of that function, I have this in the README:
```
What is nats_ReleaseThreadMemory() doing?

The NATS C library may store objects using local thread storage. Threads that are created and
handled by the library are automatically cleaned-up, however, if the user creates a thread
and invokes some NATS APIs, there is a possibility that the library stored something in that
thread local storage. When the thread exits, the user should call this function so that
the library can destroy objects that it may have stored.
```
So yes, technically if you want to make sure that the NATS C library frees resources that may have been created in a thread, you should call this function when the thread exits.

> if I call release function in thread that did not call any nats function (for example, it was stopped and did not have time to perform any work) - will it be an error?

No, this function will look for a thread local variable and if not found, won't do anything.

**Comment by kozlovic** [comment]
> can you tell me whether natsSubscription_IsValid returning false provides strong guarantees that onMsg callback will not be called more for the subscription?

No, it simply indicates that the subscription has been closed (unsubscribe or reached the auto-unsubscribe). 
If you want to guarantee that no more callback would be invoked, you should be calling it from the message callback itself. Obviously, in this case if there were pending messages (about to be dispatched), those messages are "dropped".

> can you add some callback to discover when subscription is closed and no more messages will be received. Without such a callback I have troubles with parameters passed as onMsg closure - if I create them dynamically it's hard to determine when to free them (or maybe I'm missing some possibility).

I could, I actually have one that is currently not exported/public. That being said, I thought you would be using synchronous subscription in the context of libuv, if so, that is irrelevant.

**Comment by pananton** [comment]
> That being said, I thought you would be using synchronous subscription in the context of libuv, if so, that is irrelevant.

I've decided to implement 2 modes of operation - first one named "default" (uses the same design with async subscriptions as in your library examples) and second one named "poll" (do not block waiting for events, process only pending events in single thread calling necessary callbacks and return) So I'm using async subscriptions too.

> If you want to guarantee that no more callback would be invoked, you should be calling it from the message callback itself. Obviously, in this case if there were pending messages (about to be dispatched), those messages are "dropped".

You mean that I should call  natsSubscription_Unsubscribe from onMsg callback? Btw, if I use libuv is unsibscribe performed asynchronously (i.e. it does not wait any nats protocol message from server etc)?

Also let me explain, why I encountered a problem with unsubscribe. I can dynamically subscribe/unsubscribe multiple times while application is running. I need to allocate some data from dynamic memory to be used as subscriptions onMsg closure. And of course I need to free it only when I'm sure that no more callbacks will be called. Currently I can't find easy solution to determine when to free data. With your comment about natsSubscription_Unsubscribe (if I understood correctly)  I can't understand how to do this.. Simply calling unsubscribe and then freeing allocated data is obviously bad - onMsg called later can access deallocated memory. Setting some flag indicating that subscription is closed and freeing data later in onMsg callback is also bad: subscription can be idle and receive no messages (onMsg will not be called for a long time), when onMsg is finally called and subscription is closed, some already received messages are lost (as you said). Hope I'm missing some obvious solution, will be grateful if you point it to me.

**Comment by kozlovic** [comment]
> You mean that I should call natsSubscription_Unsubscribe from onMsg callback? Btw, if I use libuv is unsibscribe performed asynchronously (i.e. it does not wait any nats protocol message from server etc)?

Yes, the idea is that if you unsubscribe from the message callback, this removes the subscription from the connection right there, so no more message will be dispatched. 
And yes, the unsubscribe sends a protocol but does not wait for a reply, so in the case of libuv, it will be just added to the buffer and libuv will flush it out.

>  I can dynamically subscribe/unsubscribe multiple times while application is running.

But if you do this out-of-band from the message callback, then you cannot expect to not "lose" messages that would be inflight. Meaning if you arbitrarily decide to call unsubscribe() without knowing if there are messages inflight for the subscriptions, you should not expect to not lose any message. What metric or event are you using to decide if you need to unsubscribe?

> Setting some flag indicating that subscription is closed and freeing data later in onMsg callback is also bad: subscription can be idle and receive no messages (onMsg will not be called for a long time)

We have options to create async subscriptions with a timeout, which means that even if there is no message, the callback is invoked after the timeout with a NULL message. See https://nats-io.github.io/nats.c/group__conn_sub_group.html#gacb2c83b58b7909715424cbc327fdd404.

> when onMsg is finally called and subscription is closed, some already received messages are lost (as you said)

But like I said, if you unsubscribe from within the message handler, then no new message will be dispatched. If you unsubscribe outside of the message handler, you are doing so already not knowing (caring) if there are inflight messages.

To summarize, we can add a "on complete" function that would be invoked when a subscription has been closed and the message handler has returned, I am just trying to see if that is the only way to solve your issue and ask you to keep in mind the facts mentioned above.

**Comment by pananton** [comment]
> But if you do this out-of-band from the message callback, then you cannot expect to not "lose" messages that would be inflight.

Totally agree. Of course, messages can be lost and it's ok for publish/subscribe model itself which should not provide any delivery guarantees.

So the problem is not about lost messages. What I'm worried about is following use case (all my future thoughts are for nats.c with libuv): application uses async subscriptions and constantly subscribes/unsubscribes to some subject (probably even the same) from distinct threads. This is some strange sort of operation, do not know where and why it can be used, but this does not matter if we provide subscribe/unsubscribe and do not want to be restrictive about their usage. When natsSubscription_Unsubscribe returns, how can app code determine when it can call natsSubscription_Destroy safely? AFAIK immediately after return from natsSubscription_Unsubscribe it is not safe, because some async onMsg can be running and can access natsSubscription. So later at some point of time we need an indication that all onMsg have finished and no more will be called. The same problem arises when we want to use natsSubscription_Drain - how to determine when all messages drained and all onMsg are called? I do not want to block in natsSubscription_WaitForDrainCompletion...

So let me summarize my questions (all for nats.c with libuv):

1. How to determine when natsSubscription can be safely destroyed when using async subscriptions and natsSubscription_Unsubscribe / natsSubscription_Drain (probably next method works)?
2. Does nats.c library guarantees that after return from natsSubscription_Unsubscribe and return from already called async onMsg callbacks (which can be called before unsubscribe and was not finished when natsSubscription_Unsubscribe was called) no more onMsg callbacks will be called, i.e. if I can determine that no onMsg are in action after natsSubscription_Unsubscribe returns it is safe to destroy natsSubscription)?
3. When I enable **echo** for connection can I be sure that published message will be received in the next example? What if `conn` is currently reconnecting? I mean, does library provide guarantee that subscription creation and publishing for the same connection will be properly ordered all the time?

`
natsConnection_Subscribe(&sub, conn, "subject", onMsg, NULL);
natsConnection_Publish(conn, "subject", "message", strlen("message"));
// will onMsg be called for sure ?
`

I have some proposals (for which I can create separate issues if you agree they are necessary):

1. Provide callback indicating that subscription is closed and no more onMsg callbacks will be called for it.
2. I'm also finding useful to have callback that indicates when subscription is created. When using libuv I have no strong guarantees: when I call natsConnection_Subscribe or natsConnection_QueueSubscribe I suppose that nats protocol message can be buffered until socket is ready to write. In my app tests I create publisher and subscriber, connect them to nats cluster, then create some subscription and start publishing to it. I think that because of race conditions some initial published messages can be lost, because subscription creation nats protocol message is delayed for some reasons. So my test will fail, because number of published messages will not be equal to number of received. Currently I use very dummy solution sleeping for some amount of time. 

Please, excuse me if I've missed some solutions that make my proposals unnecessary and you will spend time explaining this)

**Comment by kozlovic** [comment]
> When natsSubscription_Unsubscribe returns, how can app code determine when it can call natsSubscription_Destroy safely? AFAIK immediately after return from natsSubscription_Unsubscribe it is not safe, because some async onMsg can be running and can access natsSubscription.

I get your point, but actually, calling natsSubscription_Destroy() after natsSubscription_Unsubscribe() is safe, since we use reference counting internally. What is not, is freeing objects that may be used in the message handler.

> How to determine when natsSubscription can be safely destroyed when using async subscriptions and natsSubscription_Unsubscribe / natsSubscription_Drain (probably next method works at least for natsSubscription_Unsubscribe)?

Again, subscription can be destroyed and resources will be properly freed when the subscription dispatcher's thread returns, the problem is for objects used in the message handler.

> Does nats.c library guarantees that after return from natsSubscription_Unsubscribe and return from already called async onMsg callbacks (which can be called before unsubscribe and was not finished when natsSubscription_Unsubscribe was called) no more onMsg callbacks will be called, i.e. if I can determine that no onMsg are in action after natsSubscription_Unsubscribe returns it is safe to destroy natsSubscription)?

No, unless you call unsubscribe from the message handler itself. The library then remove the subscription from the connection's map and does not dispatch any of the possible remaining messages in the internal queue.

> When I enable echo for connection can I be sure that published message will be received in the next example? What if conn is currently reconnecting? I mean, does library provide guarantee that subscription creation and publishing for the same connection will be properly ordered all the time?

Note that echo is the default. Even if disconnected, the subscription will add the protocol to the internal buffer, followed by published messages. On reconnect, the subscriptions previously created are resent and the reconnect buffer is then flushed. That being said, it is possible that the reconnect buffer size is not big enough to hold all pending actions, but then you would get an error on subscribe/publish calls.

> I have some proposals (for which I can create separate issues if you agree they are necessary):
Provide callback indicating that subscription is closed and no more onMsg callbacks will be called for it.

As I said, I already have `natsSub_setOnCompleteCB()`, so I could make this a public function.

> I'm also finding useful to have callback that indicates when subscription is created.

There is no way to do that since there is no feedback from the server. BUT, you have the implicit guarantee that for the SAME connection, if a subscribe is done before the publish, then the server will process the subscribe first, so published message would be received.

If using different connections, you would need to call natsConnection_Flush() to make sure that the server has processed the subscription before starting to publish (again, from different connection). But with a cluster of server, you don't even have that guarantee if the publish connection is not on the same server.

**Comment by kozlovic** [comment]
@pananton Can you have a look at PR #303 to see if that would solve your issue?

**Comment by pananton** [comment]
@kozlovic Thanks, that's what I needed!

**Comment by pananton** [comment]
I has question again) Currently I call natsConnection_DrainTimeout and when uv loop exits I check for last error with natsConnection_GetLastError as we discussed earlier. I suppose that if returned status is NATS_OK then connection was successfully drained, otherwise - it was not. But I found some strange behaviour during stress test (20 000 messages between N publishers and subscribers) - while all messages are received by all subscribers,  natsConnection_GetLastError returns NATS_CONNECTION_CLOSED when some subscribers are stopped. I debugged nats.c a little and found that this error is set because socket read returns 0 (meaning that nats server initiates tcp shutdown). This happens very rare in my test (about 1 out of 100 times). Maybe you know what can be a reason? I've tried to launch nats server with trace logs, but to reproduce this "error" I need big number of messages and subscribers and in that case nats logs become large and error disappears (probably because flushing to disk takes time from nats server and client closes connection faster). Maybe you have any ideas what can be the reason?

**Comment by kozlovic** [comment]
The server should not initiate a tcp shutdown unless the client is sending bad protocol or there is a perm error, etc.. My guess is that there is a race between closing the connection and a read that then returns that the connection is closed. But looking at the code, it looks like we would handle this condition properly detecting that the connection has been marked as closed and not update the last error.

Where do you see the read error setting the last error to NATS_CONNECTION_CLOSED? The low level natsSock_Read() returns NATS_CONNECTION_CLOSED if number of bytes read is 0, but that does not set the connection's last error. natsConnection_ProcessReadEvent() does call natsSock_Read() and on error calls _processOpError() which then do nothing if it detects that the connection has been closed.

I would need a reproducible test case to investigate further.

**Comment by pananton** [comment]
I've tried to create a test using only nats.c library functions (stripping off my own functionality) but can't reproduce exactly the same situation. Instead, I've found another issue. Can you please look into to example and tell me, whether code looks correct or I'm missing something? I've added this function to your test.c file and run it several times. And I'm getting a timeout error for natsConnection_DrainTimeout from time to time. Probably I completely misunderstand how lib works.. 
```
typedef struct {
    natsConnection *conn;
    natsPid serverPid;
} drainThreadInfo;

static void drainThread(void *arg)
{
    printf("start draining\n");
    drainThreadInfo *info = (drainThreadInfo*)arg;
    natsStatus s = natsConnection_DrainTimeout(info->conn, 1000);
    printf("natsConnection_DrainTimeout: %d\n", (int)s);
}

static void
test_natsConnStatusAfterReconnectAndStop(void)
{
    natsPid     serverPid = NATS_INVALID_PID;
    drainThreadInfo info;
    info.serverPid = _startServer("nats://127.0.0.1:4333", "-p 4333", true);

    natsStatus  s;
    natsOptions *opts = NULL;
    natsConnection *conn = NULL;

    natsLibuv_Init();

    s = natsOptions_Create(&opts);
    natsOptions_SetURL(opts, "nats://127.0.0.1:4333");
    natsOptions_SetAllowReconnect(opts, true);
    natsOptions_SetMaxReconnect(opts, 0x70000000);
    natsOptions_SetReconnectWait(opts, 200);

    uv_loop_t *uv = (uv_loop_t*)(malloc(sizeof(uv_loop_t)));
    uv_loop_init(uv);
    uv_thread_t thr;

    natsLibuv_SetThreadLocalLoop(uv);
    natsOptions_SetEventLoop(opts, uv, natsLibuv_Attach, natsLibuv_Read, natsLibuv_Write, natsLibuv_Detach);

    s = natsConnection_Connect(&conn, opts);
    info.conn = conn;

    while (uv_run(uv, UV_RUN_ONCE) && natsConnection_Status(conn) != NATS_CONN_STATUS_CONNECTED)
    {
        printf("waiting for connect 1\n");
        nats_Sleep(20);
    }

    printf("connected \n");
   _stopServer(info.serverPid);

    while (uv_run(uv, UV_RUN_ONCE) && !natsConnection_IsReconnecting(conn))
    {
        printf("waiting for disconnect\n");
        nats_Sleep(20);
    }

    printf("reconnecting\n");
    info.serverPid = _startServer("nats://127.0.0.1:4333", "-p 4333", true);

    while (uv_run(uv, UV_RUN_ONCE) && natsConnection_Status(conn) != NATS_CONN_STATUS_CONNECTED)
    {
        printf("waiting for connect 2\n");
        nats_Sleep(20);
    }

    printf("connected 2\n");

    // starting connection drain
    uv_thread_create(&thr, drainThread, (void*)&info);

    // waiting until drain completes
    while (uv_run(uv, UV_RUN_ONCE)) { nats_Sleep(20); }
    printf("uv loop exited\n");
    uv_thread_join(&thr);
    printf("drain thread joined\n");

    const char *errorString;
    s = natsConnection_GetLastError(conn, &errorString);

    if (s != NATS_OK)
    {
        printf("UNEXPECTED: %d\n", (int)s);
    }

    if (uv)
    {
        uv_loop_close(uv);
        free(uv);
    }

    if (conn)
    {
        natsConnection_Destroy(conn);
        conn = NULL;
    }

    if (opts)
    {
        natsOptions_Destroy(opts);
        opts = NULL;
    }

    _stopServer(info.serverPid);
    info.serverPid = NATS_INVALID_PID;
    printf("all done\n");
}

==============
OUTPUT IS :

connected
reconnecting
connected 2
start draining
 // here test block for about 1 minute
natsConnection_DrainTimeout: 0
uv loop exited
drain thread joined
UNEXPECTED: 26
all done
ALL PASSED
```

**Comment by pananton** [comment]
Also I'd like to ask you, how nats.c manages last error info? I use GetLastError to determine whether connection was closed (expect that last error will always be nats_ok in that case) or lost (will never be nats_ok in that case). I doubt that this will work when reconnect is enabled, because I suppose that after connection is restored last error can stay the same until first successful operation (for example, publish). Is it true? Or last error will change to nats_ok right after reconnect? Maybe I should use natsConnStatus to determine whether connection is closed normally or was lost?

**Comment by kozlovic** [comment]
@pananton Thanks for the test, I have been able to reproduce (not that often, which makes it more difficult, but still able to reproduce). I spent most of today trying to understand and I am making some progress. I will keep you updated.

> Maybe I should use natsConnStatus to determine whether connection is closed normally or was lost?

You can setup callbacks to be notified if the connection was disconnected and/or when getting errors (natsOptions_SetDisconnectedCB, natsOptions_SetErrorHandler, etc..). That being said, unless you are explicitly calling natsConnection_Close/Destroy, you should be able to tell if the connection is closed because of error on the connection that caused it to close.

**Comment by kozlovic** [comment]
@pananton After much debugging, I have realized that the issue is with the test case, not the library. You need to replace all your use of `UV_RUN_ONCE` with `UV_RUN_NOWAIT`, and you should have this to wait for the drain to complete (again in this test):
```
    // waiting until drain completes
    while (uv_run(uv, UV_RUN_NOWAIT) && !natsConnection_IsClosed(conn))
    {         
        printf("waiting for drain to timeout\n");
        nats_Sleep(20); 
    }
```
Why NOWAIT instead of ONCE? From the doc (http://docs.libuv.org/en/v1.x/loop.html):
```
UV_RUN_ONCE: Poll for i/o once. Note that this function blocks if there are no pending callbacks. Returns zero when done (no active handles or requests left), or non-zero if more callbacks are expected (meaning you should run the event loop again sometime in the future).
```
That means that for instance when you were trying to see if the connection was connected, if the status was not CONNECTED when reaching this code:

**Comment by kozlovic** [comment]
> I may have an update to the libuv.h to schedule the attach if the list is not empty (even if attach is invoked from libuv thread) to ensure ordering of events.

Actually there is no need because a reconnect will ALWAYS call attach from a different thread than the libuv thread, so it will always be scheduled.

**Comment by pananton** [comment]
@kozlovic Ivan, sorry that you spent time for this. I've read libuv documentation earlier and forgot about this. Also UV_RUN_ONCE sounds like something that I need, so I did not check it in documentation, my bad. Thank you for help.

**Comment by pananton** [comment]
Hello, Ivan. Can you please explain, why it is not allowed to use several uv loops in one thread in nats.c? Can this be easily changed? For example, the following code fragment does not work (executed from the same thread):
```
    ...
    natsLibuv_SetThreadLocalLoop(uv1);
    natsOptions_SetEventLoop(opts1, uv1, natsLibuv_Attach, natsLibuv_Read, natsLibuv_Write, natsLibuv_Detach);

    natsLibuv_SetThreadLocalLoop(uv2);
    natsOptions_SetEventLoop(opts2, uv2, natsLibuv_Attach, natsLibuv_Read, natsLibuv_Write, natsLibuv_Detach);

    s = natsConnection_Connect(&conn1, opts1);
    testCond(s == NATS_OK);
    ...

============
OUTPUT:
Error: 19 - Illegal State - (conn.c:1864): Error attaching to the event loop: 19 - Illegal State
Stack: (library version: 2.1.0)
  01 - _processConnInit
  02 - _connect
  03 - natsConnection_Connect
*** 1 TEST FAILED ***
```

Such usage of libuv can be useful when using UV_RUN_ONCE / UV_RUN_NOWAIT

**Comment by kozlovic** [comment]
> Can you please explain, why it is not allowed to use several uv loops in one thread in nats.c?

Because libuv is not thread safe, so we need to know if a call to libuv is done from the event loop thread or not. That is the point of `natsLibuv_SetThreadLocalLoop()` which registers the loop into the current thread. NATS libuv adapter then checks that its loop is find as a thread local in the current thread and if not, need to schedule it.

If we could step back a bit, what is your use case so we can help you better. I have been answering low level questions, but maybe we should have a better idea of the bigger picture.

For instance, the direction of having multiple event loops on the same thread seem strange to me. What one would want to do that? On the contrary, people would use event loop with NATS C client to actually reduce the number of threads (since by default we use a thread per connection to read data) and "bundle" all the events on the same event loop.

**Comment by pananton** [comment]
Ok, let me explain ideas behind my design. I'm writing a C++ wrapper library for nats.c to be used in our microservices communicating with nats. I prefer to use libuv not only because it allows to reduce number of threads (which also can be done using thread pool afaik) but also because it makes nats.c true async (we've discussed earlier that publish call can block without libuv because underlying socket send will block when network congested). Second, in existing evironment we already using a very simple (but a little inefficient) threading model - thread sleeps for a fixed (or auto tuned) interval (10-20 ms) and then epolls opened sockets (parsing our proprietarya protocol) and other external libraries (for example, grpc that provides method on it's channel interface to poll for ready events). Because all is performed in one thread and all callbacks also get called on one thread we can avoid synchronization at all - whether event comes from our socket, grpc, some db handler. For this reason my library provides a way to incorporate nats.c underlying event loop to other thread (i.e., make it work in a specified thread without blocking it). For example, like this:
```
void runService() // service thread function
{
    while (!isStopped_) {
        // each `run` calls callbacks which operate on shared data but no
        // synchronization is required.

        grpc->run(); // process all pending grpc calls, deliver pending requests, and return
        dbHandler->run(); // process all received from DB data, send ready requests and return
        nats->run(); // process all pending nats events, deliver ready data and return
        sleep(20);
    }
}
```
So this is my motivation to use uv loop in UV_RUN_NOWAIT mode.
Next, because libuv is not used in our projects I do not want to make it public dependency of my lib and it is used only internally. Main interface is `Client` which provides methods to publishing/subscribing and `run` which does actual work (basically - calls uv_run in selected mode (UV_RUN_DEFAULT / UV_RUN_NOWAIT)). Each `Client` has own natsOptions, incorporates single connection and single uv loop. I can create several clients in one app if I want to configure them differently (for example, with/without asap flag). If all `Client` instances use UV_RUN_NOWAIT mode then nothing prevents from running them on one thread from libuv's point of view (at least, here it is stated that it's ok for libuv, was not able to find more information: http://nikhilm.github.io/uvbook/multiple.html). But looks like nats.c does not allow this because natsLibuv_SetThreadLocalLoop called for distinct loops on ONE thread brokes something. Probably I should move loop management from `Client` to separate class and make `Client` instances attach to it, need to think more how this can be done.

**Comment by pananton** [comment]
I've started refactoring my code and have a question: does nats.c with libuv allows to attach several connections to one uv_loop_t? I mean, will the following code work:
```
// uv is a pointer to initialized uv_loop_t
natsLibuv_SetThreadLocalLoop(uv);
natsOptions_SetEventLoop(options1, uv, ...);
natsOptions_SetEventLoop(options2, uv, ...); // same uv loop
natsConnection_Connect(&conn1, options1);
natsConnection_Connect(&conn2, options2);
uv_run(uv);
```

If this will work, will uv_run return only when all nats connections are closed?

**Comment by kozlovic** [comment]
> does nats.c with libuv allows to attach several connections to one uv_loop_t? 

Yes.

> If this will work, will uv_run return only when all nats connections are closed?

Yes.

**Comment by pananton** [comment]
Hello! I've finally added callback for unsubscribed subscriptions to my code and have a small question. When I have this cb set and also cb for connection close (set with natsOptions_SetClosedCB) will the former always be called for all active subscriptions before the last one? Or maybe they can interleave ?

**Comment by kozlovic** [comment]
They are independent, meaning that the subscription "on complete" callback will be invoked from the subscription's dispatcher thread (or the library messages dispatcher thread pool), while the connection closed CB will be invoked from the library async CB dispatcher thread.
From a pure NATS objects, it does not matter because we use reference counting, so even if you were to call natsConnection_Destroy() before the last subscription "on complete" callback fires, the subscription's dispatcher thread has a reference to the connection and will decrement it only on exit, possibly freeing the connection object there.

**Comment by pananton** [comment]
Thanks for clarification!

**Comment by kozlovic** [comment]
Closing for now, feel free to re-open if you have more questions or create a new issue if applicable.
